src/test/resources/com/cloudbees/plugins/credentials/casc/credentialsProviderManager.yaml

```diff
@@ -0,0 +1,14 @@
+globalCredentialsConfiguration:
+  configuration:
+    providerFilter:
+      excludes:
+        classNames:
+          - "com.cloudbees.plugins.credentials.UserCredentialsProvider"
+    restrictions:
+      - includes:
+          provider: "com.cloudbees.hudson.plugins.folder.properties.FolderCredentialsProvider"
+          type: "com.cloudbees.plugins.credentials.impl.UsernamePasswordCredentialsImpl$DescriptorImpl"
```

Comment:
Hmm, nothing in JCasC should ever be referring to a `Descriptor` class name. It is not even required that you implement `Descriptor` with a distinct type—you can use a static field or an anonymous inner class or whatever. While we may be required to allow such an entry in a historical `config.xml`, going forward, we should be using `Descriptor.getId()`, which is _usually_ (but not necessarily) a `Describable` class name.

Reply:
any psuedo-code adapting hint? or an example?

Reply:
Do you mean for the `readResolve` impl, or for using `Descriptor.id` going forward?

Reply:
descriptor.id

Reply:
do you mean just store the descriptor id instead of the full class name?

Reply:
Right, so in this example `com.cloudbees.plugins.credentials.impl.UsernamePasswordCredentialsImpl` rather than `com.cloudbees.plugins.credentials.impl.UsernamePasswordCredentialsImpl$DescriptorImpl`.

Reply:
or the symbol if it has one? `usernamePassword` ? I assume can be looked up pretty easily

Reply:
The UI is built off of https://github.com/jenkinsci/credentials-plugin/blob/master/src/main/resources/com/cloudbees/plugins/credentials/CredentialsProviderManager/Configuration/config.jelly#L30

Reply:
not sure much how it works need to track it back

Reply:
> or the symbol if it has one?

If you like.

Reply:
Hmm I'm not sure if this functionality is broken or just only affects the UI.

I've made the changes locally (although no backwards compat added yet).

From my testing the only thing that was affected was the doFill credentials from credentials binding. (probably anything using the credentials jelly control) Running a job where you'd already selected the credential and then restricted it works just fine. (Or using pipeline and free text filling it in)

Anyone know if that's intended?

Docs appear to be here: https://github.com/jenkinsci/credentials-plugin/blob/master/docs/user.adoc
Although there's no inline help

Reply:
Yes that's intended. The include and exclude filters are there for making the UI easier to use. Permissions are what control usage restrictions.

Reply:
Include and exclude work how I expect and prevent them being used, restrictions is the weird one but ok if it’s UI only…

Reply:
Something weird going on with descriptor's not being loaded =/, https://github.com/jenkinsci/credentials-plugin/pull/235/files#r721098631

Any ideas?
Otherwise maybe drop the test? Or even the `readResolve` given this seems to be a UI only thing.

Reply:
Actually I'll just use the dummy credentials as that seems to be what other tests do, still weird it doesn't work